### PR TITLE
Add tests for frames

### PR DESCRIPTION
**Pull request by Granitosaurus**

### CODE REVIEW

File: tests/test_frame.py [hunk 1]
```diff
@@ -1,640 +1,183 @@
...
+        context1.evaluate('window.a'),
+        context2.evaluate('window.a'),
+    )
+    assert a1 == 1
+    assert a2 == 2
+
+@sync
+async def test_evaluate(isolated_page, server):
+    p = isolated_page
+    # should throw for detached frames
```
Comment (Mattwmaster58) [inline tests/test_frame.py]: In general I'd prefer more descriptive test fn names by default (and I think most of the tests in this file are appropriately named). This name will provide almost no information when it fails.
```suggestion
async def test_evaluate_raises_on_detaches_frames(isolated_page, server):
    p = isolated_page
```

File: tests/test_frame.py [hunk 2]
```diff
@@ -1,640 +1,183 @@
...
+    )
+    assert p.url == server.empty_page + '#foo'
+    # should persist mainFrame on cross-process navigation
+    await p.goto(server.empty_page)
+    mainFrame = p.mainFrame
+    await p.goto(server.cross_process_server / 'empty.html')
+    assert p.mainFrame is mainFrame
+
+@sync
+async def test_attaching(isolated_page, server):
```
Comment (Mattwmaster58) [inline tests/test_frame.py:92]: Same issue mentioned above

File: tests/test_frame.py [hunk 3]
```diff
@@ -1,640 +1,183 @@
...
+    assert len(detachedFrames) == 0
+    assert len(navigatedFrames) == 5
+
+    # should detach child frames on navigation
+    attachedFrames = []
+    detachedFrames = []
+    navigatedFrames = []
+    await p.goto(server.empty_page)
+    assert len(attachedFrames) == 0
+    # TODO here detachedframes has more than 4 because it's filled with dupes
```
Comment (Mattwmaster58) [inline tests/test_frame.py]: I'm assuming it doesn't anymore cause this test passes 😉

File: tests/test_frame.py [hunk 4]
```diff
@@ -1,640 +1,183 @@
...
+    assert len(detachedFrames) == 4
+    assert len(navigatedFrames) == 1
+
+@sync
+async def test_report_frame(isolated_page, server):
+    p = isolated_page
+    # should report frame from-inside shadow DOM
```
Comment (Mattwmaster58) [inline tests/test_frame.py]: Function naming, mentioned earlier

File: tests/test_frame.py [hunk 5]
```diff
@@ -1,640 +1,183 @@
...
+    frame1 = await attachFrame(p, server.empty_page, 'frame1')
+    await p.evaluate(
+        """
+        () => {
+            window.frame = document.querySelector('#frame1');
+            window.frame.remove();
+        }
+        """
+    )
+    assert frame1.isDetached is True
```
Comment (Mattwmaster58) [inline tests/test_frame.py]: Nitpick, but more idiomatically expressed this could be `assert frame1.isDetached`

**Review by Mattwmaster58 — CHANGES_REQUESTED** [review]

File: tests/test_frame.py [hunk 6]
```diff
@@ -1,640 +1,183 @@
...
+
+@sync
+async def test_evaluate(isolated_page, server):
+    p = isolated_page
+    # should throw for detached frames
+    frame1 = await attachFrame(p, server.empty_page, 'frame1')
+    await detachFrame(p, 'frame1')
+    with pytest.raises(BrowserError) as e:
+        await frame1.evaluate('7 * 8')
+    assert e.match('Execution Context is not available in detached frame')
```
Comment (Mattwmaster58) [inline tests/test_frame.py]: IMO the inline exception message matching is a little better:
```suggestion
    with pytest.raises(BrowserError, match='Execution Context is not available in detached frame') as e:
        await frame1.evaluate('7 * 8')
```

Goes for other examples too, I'm just tool lazy to highlight them all 😛

**Comment by Granitosaurus** [comment]
@Mattwmaster58 I resolve the comments and some tests were failing because of the fact that python sets are unordered while javascript's are. 

I think it would solve us a lot of headache to migrate sets to orderedsets as we can't really know when the importance of ordering could pop up.  
I've added OrderedSet dependancy in latest commit.

File: pyppeteer/frame/__init__.py [hunk 7]
```diff
@@ -31,7 +32,7 @@ def __init__(
         self._lifecycleEvents: Set[str] = set()
         self._mainWorld = DOMWorld(frameManager, self, frameManager._timeoutSettings)
         self._secondaryWorld = DOMWorld(frameManager, self, frameManager._timeoutSettings)
-        self._childFrames: Set[Frame] = set()  # maybe list
+        self._childFrames: OrderedSet[Frame] = OrderedSet()
```
Comment (Mattwmaster58) [inline pyppeteer/frame/__init__.py:36]: There was an issue on the orderedset project asking about type hinting, and it implied this currently isn't possible. I think it's fine to keep it in for now though, there's hundreds of mypy errots and we can easily change this one in a batch of mypy fixes.

**Review by Mattwmaster58 — APPROVED** [review]

**Comment by adrianvanlan** [comment]
Hello, first of all excellent work!
I wanted to know if you are thinking about "setContent".
Reference: https://github.com/puppeteer/puppeteer/blob/main/src/FrameManager.ts#L477-L485

**Comment by Mattwmaster58** [comment]
> I wanted to know if you are thinking about "setContent".

That's already implemented in the branch this PR merged too.

**Pull request merged**